### PR TITLE
Add NVTX ranges to categorize execution

### PR DESCRIPTION
**Pull request by minitu**
# What does this PR do ?

Adds NVTX ranges inside NeMo to categorize different parts of the execution.
For now the PR only adds an NVTX range around the MCore optimizer step.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by minitu** [comment]
@gautham-kollu

File: nemo/utils/nvtx.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import functools
+import torch
+
+@functools.lru_cache
+def _nvtx_enabled() -> bool:
+    return bool(int(os.getenv("NEMO_NVTX_ENABLED", "0")))
```
Comment (minitu) [inline nemo/utils/nvtx.py:21]: Needs more discussion on how this should be controlled.

Comment (gautham-kollu) [reply nemo/utils/nvtx.py:23]: @marcromeyn  Ideally we want this to be part of the Nsyscallback().  

What is a good way to read this config is enabled ? Both the model and the optimizer would interested to know is_nvtx_range() enabled ?

cc: @hemildesai

**Review by gautham-kollu — APPROVED** [review]

File: nemo/core/optim/mcore_optim.py [hunk 2]
```diff
@@ -42,6 +44,8 @@ class McoreDistributedOptimizer(torch.optim.Optimizer):
         optim (MegatronOptimizer): The distributed optimizer from Megatron Core.
     """
 
+    nvtx_label = "nemo.core.optim.mcore_optim"
```
Comment (gautham-kollu) [inline nemo/core/optim/mcore_optim.py:47]: np: NVTX_LABEL =

Comment (minitu) [reply nemo/core/optim/mcore_optim.py]: Updated

**Review by gautham-kollu — APPROVED** [review]

**Review by gautham-kollu — APPROVED** [review]
LGTM
